src/pudl/analysis/demand_mapping.py

```diff
@@ -688,3 +742,492 @@ def categorize_eia_code(eia_codes, ba_ids, util_ids, priority="balancing_authori
+        total_geom (geopandas.GeoDataFrame): This is a geodataframe which has
+            the total US mainland map. This will allow display of the regions
+            not covered even once as grey.
+        title (str): The title for the map visualization
+
+    Returns:
+        None: Displays the image
+
+    """
+    covered_geom = (disagg_geom[["respondent_id_ferc714_set", "geometry"]]
+                    .drop_duplicates()
+                    .reset_index(drop=True)
+                    )
+    covered_geom["num_covered"] = covered_geom["respondent_id_ferc714_set"].apply(
+        len)
+    max_val = covered_geom["num_covered"].max()
+
+    # extract viridis features
+    cmap = plt.cm.viridis  # define the colormap
+    cmaplist = [cmap(i) for i in range(cmap.N)]
+    # force the first color entry to be grey
+    cmaplist[0] = (.5, .5, .5, 1.0)
+    # create the new map
+    cmap = mpl.colors.LinearSegmentedColormap.from_list(
+        'DiscreteViridis', cmaplist, cmap.N)
+
+    # define the bins and normalize
+    bounds = np.linspace(-0.5, max_val + 0.5, max_val + 2)
+    ticks = [b + 0.5 for b in bounds][:-1]
+    norm = mpl.colors.BoundaryNorm(bounds, cmap.N)
+
+    # create the figure
+    fig, ax = plt.subplots(figsize=(20, 13))
+    ax = total_geom.assign(num=1).dissolve(
+        by="num").plot(facecolor="grey", ax=ax)
+    covered_geom.plot("num_covered", legend=False, ax=ax, cmap=cmap, norm=norm)
+    plt.title(title, fontdict={'fontsize': 25})
+
+    ax.set_xticks([])
+    ax.set_yticks([])
+
+    # create a second axes for the colorbar
+    ax2 = fig.add_axes([0.97, 0.1, 0.03, 0.8])
+    mpl.colorbar.ColorbarBase(ax2, cmap=cmap, norm=norm,
+                              spacing='proportional', ticks=ticks,
+                              boundaries=bounds, format='%1i')
+
+    plt.show()
+
+
+def vec_error(vec1, vec2, errtype):
+    """
+    Calculate error metrics between two vectors vec1 (alloc), and vec2 (actual).
+
+    Takes input of two numpy arrays, vec1 and vec2, and calculates error metric.
+    Possible specifications of error metrics include: Mean Squared Error
+    ('mse'), Mean Absolute Percentage Error ('mape%') and R2 value ('r2').
+    """
+    vec1 = np.array(vec1)
+    vec2 = np.array(vec2)
+    if errtype == "mse":
+        return np.nanmean((vec1 - vec2) ** 2)
+
+    elif errtype == "mape%":
+        vec1[vec2 == 0] = np.nan
+        return np.nanmean(np.abs((vec1 - vec2) / vec2)) * 100
+
+    elif errtype == "r2":
+        mask = ~np.isnan(vec1) & ~np.isnan(vec2)
+        if vec1[mask].size == 0:
+            return np.nan
+
+        else:
+            _, _, r_value, _, _ = scipy.stats.linregress(
+                vec1[mask], vec2[mask])
+            return r_value ** 2
+
+
+def error_heatmap(alloc_df, actual_df, demand_columns, region_col="pca", error_metric="r2", leap_exception=False):
+    """
+    Create heatmap of 365X24 dimension to visualize the annual hourly error.
```

Comment:
This function does a very particular thing, and I think it might be better separated into a few smaller functions that are more reusable -- like one which, given a year of hourly data (a Datetime index + a data column) makes a heatmap, and another separate one which takes 2 hourly demand allocations and calculates the delta between them, the output for which can be fed into the plotting function. That we we can make other 365 x 24 heatmaps to show other kinds of variables that make sense that way too.

Visually, can we scale it down so it fits on one screen? I think the Y-axis can also be simplified to have day-of-year (integer) as the labels. Could label every 7 days to highlight the weekly pattern? Could just have the first letters of the months JFMAMJJASOND.  Allowing the UTC time to be localized for display so that the zero hour is local zero would also be good, since it would make the familiar diurnal pattern clearer, and allow uniform visual comparison between different plots of this type.